### PR TITLE
Implement MAE loss function

### LINKED ISSUE #1333: MAE loss function
### Describe the task. It can be a feature, documentation, etc.

Implement MAE loss function (with weights for physical loss such as `mse_channel_location_weighted`).

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### LINKED ISSUE #1536: Implicit loss function conversion in LossPhysical
### What happened?

We currently have the mse is changed mse_channel_location_weighted implicitly here:

https://github.com/ecmwf/WeatherGenerator/blob/dfcf3e16754e2af518dd6e76e9a9ccfb7564ec41/src/weathergen/train/loss_modules/loss_module_physical.py#L52

These implicit things are hard to understand from the config. Does mse_channel_location_weighted default to an unweighted mse?

### What are the steps to reproduce the bug?

_No response_

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### PR DESCRIPTION
**Pull request by Jubeku**
## Description

Implement MAE loss function (with weights for physical loss such as mse_channel_location_weighted).

UPDATE:
A generalized lp-norm function is implemented based on which MAE, MSE, RMSE, SSE can be implemented.

## Issue Number

Closes #1333 
Closes #1536 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 1]
```diff
@@ -186,6 +186,66 @@ def mse_channel_location_weighted(
...
+
+    mask_nan = ~torch.isnan(target)
+    pred = pred[0] if pred.shape[0] == 0 else pred.mean(0)
+
+    diff_abs = torch.abs(torch.where(mask_nan, target, 0) - torch.where(mask_nan, pred, 0))
+    if weights_points is not None:
+        diff_abs = (diff_abs.transpose(1, 0) * weights_points).transpose(1, 0)
+    loss_chs = diff_abs.mean(0)
+    loss = torch.mean(loss_chs * weights_channels if weights_channels is not None else loss_chs)
+
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py:268]: If we use the implementation below (didn't run it, so needs testing) then we could support MAE, MSE, RMSE and any othr l_p norm (for p < inf) in one function. Maybe it should be called `generalized_lp_norm` but I could live with the current naming. MAE, MSE would still exist as functions but they all use lp_norm in the implementation.

```
def lp_loss( 
    target: torch.Tensor,
    pred: torch.Tensor,
    p_norm: int,
    with_p_root: bool = True,
    with_mean = True,
    weights_channels: torch.Tensor | None = None,
    weights_points: torch.Tensor | None = None,
):

assert type(p_norm) is int, "Only integer p supported for p-norm loss"

mask_nan = ~torch.isnan(target)
pred = pred[0] if pred.shape[0] == 0 else pred.mean(0)

diff_p = torch.pow( torch.abs(torch.where(mask_nan, target, 0) - torch.where(mask_nan, pred, 0)), p_norm)
if weights_points is not None:
     diff_p = (diff_p.transpose(1, 0) * weights_points).transpose(1, 0)
loss_chs = diff_p.mean(0) if with_mean else diff_p.sum(0)
loss_chs = torch.pow( loss_chs, 1.0/p) if with_p_root else loss_chs 
loss = torch.mean(loss_chs * weights_channels if weights_channels is not None else loss_chs)

Comment (tjhunter) [reply src/weathergen/train/loss_modules/loss_functions.py:268]: are you sure you want to recommend that? most of these functions have accelerated code/intrinsics where the performance difference can be shocking, especially for flowing gradient computations. I would not trust the torch compiler to detect and optimize these patterns.

**Comment by tjhunter** [comment]
@Jubeku  noting there is a conflict

**Comment by Jubeku** [comment]
This PR makes PR #1551 obsolete which removes the implicit loss function conversion in LossPhysical.

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 2]
```diff
@@ -127,18 +123,27 @@ def kernel_crps(
...
+    weights_channels: torch.Tensor | None = None,
+    weights_points: torch.Tensor | None = None,
 ):
     """
-    Compute weighted MSE loss for one window or step
+    This function computes the Lp-norm for any arbitrary integer p < inf.
+    By default, the Lp-norm is normalized by the number of samples (i.e. with_mean=True).
+    * For example: p=1 corresponds to MAE; p=2 corresponds to MSE.
+    The samples are weighted by location if weights_points is not None.
+    The norm can optionally normalised by the pth root.
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py]: optionally normalized -> optionally be normalized

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 3]
```diff
@@ -127,18 +123,27 @@ def kernel_crps(
...
+    * For example: p=2 and with_p_root=True corresponds to RMSE.
+    The mean across all channels can optionally be weighted by channel weights.
 
     The function implements:
 
-    loss = Mean_{channels}( weight_channels * Mean_{data_pts}( (target - pred) * weights_points ))
+    loss = Mean_{channels}(weight_channels * Mean_{data_pts}(|(target - pred)|**p * weights_points))
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py]: The p-norm usually involves the p-th root "on the outside". This should also be an optional argument. Please make sure it's at the right position with respect to the mean operations.

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 4]
```diff
@@ -159,92 +164,108 @@ def mse_channel_location_weighted(
...
+    3. weight the collapsed cols by wc = weights_channels (if given)
     4. take the mean over the channel-weighted cols
 
     Params:
-        target : shape ( num_data_points , num_channels )
-        target : shape ( ens_dim , num_data_points , num_channels)
-        weights_channels : shape = (num_channels,)
-        weights_points : shape = (num_data_points)
+        target : tensor of shape ( num_data_points , num_channels )
+        target : tensor of shape ( ens_dim , num_data_points , num_channels)
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py]: pred : ...

File: src/weathergen/train/loss_modules/loss_functions.py (hunk 3 above)
Comment (clessig) [reply src/weathergen/train/loss_modules/loss_functions.py]: It's actually present in the implementation but should be mentioned here then as well.

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 5]
```diff
@@ -159,92 +164,108 @@ def mse_channel_location_weighted(
...
+        p_norm : integer defining the p the type of the norm
+        with_mean : boolean defining whether the norm is summed or averaged
+        with_p_root : boolean defining whether the p-th root of the norm is returned
+        weights_channels (optional): tensor of shape = (num_channels,)
+        weights_points (optional): tensor of shape = (num_data_points)
 
     Return:
-        loss : weight loss for gradient computation
-        loss_chs : losses per channel with location weighting but no channel weighting
+        loss : (weighted) loss for gradient computation
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py]: loss: (weighted) scalar loss (e.g. for gradient computation)

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 6]
```diff
@@ -159,92 +164,108 @@ def mse_channel_location_weighted(
...
-        |                      |    |  |
-        |                      |    |  |
-        ------------------------     -
-                    x
-        ------------------------
-        |          wc          |
-        ------------------------
-
-    where wp = weights_points and wc = weights_channels and "x" denotes row/col-wise multiplication.
+    return lp_loss(
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py:216]: We should refer to lp_loss for detailed explanation of arguments.

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 7]
```diff
@@ -159,92 +164,108 @@ def mse_channel_location_weighted(
...
-    2. take the mean over the row
-    3. weight the collapsed cols by wc = weights_channels
-    4. take the mean over the channel-weighted cols
 
-    Params:
-        target : shape ( num_data_points , num_channels )
-        target : shape ( ens_dim , num_data_points , num_channels)
-        weights_channels : shape = (num_channels,)
-        weights_points : shape = (num_data_points)
+def sse(
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py]: Is "sse" a standard name? I am not familar with it.

**Review by clessig — APPROVED** [review]
Just some minor points on the documentation/comments. The code has been tested?

File: src/weathergen/train/loss_modules/loss_functions.py (hunk 7 above)
Comment (Jubeku) [reply src/weathergen/train/loss_modules/loss_functions.py]: Maybe better Residual Sum of Squares (RSS)?
https://en.wikipedia.org/wiki/Residual_sum_of_squares

**Comment by Jubeku** [comment]
> Just some minor points on the documentation/comments. The code has been tested?

I run the integration test and tested manually that the value of the mse loss corresponds to the value using the old implementation (`mse_channel_location_weighted`):
image
